### DISCUSSION TITLE
Multipart file encoding takes up a lot of memory

### DISCUSSION THREAD

**Discussion by mfleader** [encode/httpx#1659] [Potential Issue]
Using httpx AsyncClient to post a file, takes up a lot of memory. Specifically posting a 10 GiB file caused the python process to go up to 20 GiB RAM for several seconds, and posting a 20 GiB file caused the python process to consume all of my available memory ~23 GiB for much longer.

**Steps to reproduce the issue:**

1. Create files for a given size

```shell
truncate -s 5M ~/5M.txt
truncate -s 10G ~/10G.txt
truncate -s 20G ~/20G.txt
```

2. Post a file

```python
from pathlib import Path
import httpx

async def async_post_file_req(
    url: str, token: str, filepath: Path, filedir: str = ''):    
    
    async with httpx.AsyncClient(
        timeout=httpx.Timeout(write=None, read=None, 
            connect=None, pool=None)) as client:

        r = await client.post(
            url, 
            params={
                'filename': filepath.name,
                'filedir': filedir
            },
            files={
                'file': (filepath.name, filepath.open('rb'), 'application/octet-stream')
            },
            headers = {
                'Authorization': f'Bearer {token}'
            }
        )
```

3. Watch system processes
```shell
top
```

4. Make post
```python
import os

url = os.getenv('FILE_SERVER_URL')
tkn = token(f'{url}/auth/jwt/login', 
    os.getenv('FILE_SERVER_UNAME'), 
    os.getenv('FILE_SERVER_PASS'))

asyncio.run(
    async_post_file_req(
        f'{url}/stream', tkn,
        Path('/home/mleader/10G.txt')
    )
)
```

**Comment by lovelydinosaur** [comment]
That's surprising, since we do use incremental upload for multipart files... https://github.com/encode/httpx/blob/776dbb578b1fd69f84c5bf75a80cd74dc823a41c/httpx/_multipart.py#L136

Probably worth looking into some more.
First up: is anyone able to reproduce the behaviour described here?
